Transactions features

Dm1triiSmirnov opened the pull request:
Added:         
1. Validation for check that the sender and receiver wallets have the same currency. Transactions are available only for wallets with the same currency
2. Commission calculation based on whether the sender and receiver wallets belong to the same user. When user sends money from his wallet to his another wallet - no commission, and when he sends to wallet, related to another user - commission=10%

Dm1triiSmirnov committed:
add features for transactions: 1. validation for transaction currency; 2. calculations the commission

VladKli commented on transaction/models.py:
```diff
@@ -16,7 +16,7 @@ class Transaction(models.Model):
-    commission = models.FloatField()
-    status = models.CharField(choices=STATUS, max_length=100)
+    transfer_amount = models.DecimalField(max_digits=12, decimal_places=2)
+    commission = models.DecimalField(max_digits=12, decimal_places=2, default=0)
+    status = models.CharField(choices=STATUS, max_length=100, default='PAID')
     timestamp = models.DateTimeField(auto_now_add=True)
```
please add `editable` attr to the `commission`, `status` and `timestamp` fields and set it to `False`
also add default value for the `transfer_amount` equals to 0

Dm1triiSmirnov replied:
Done

Dm1triiSmirnov committed:
add check for sufficient amount of sender balance

Dm1triiSmirnov committed:
fix transaction/models.py add editable=False to some fields

Dm1triiSmirnov committed:
resolve conflicts

Dm1triiSmirnov committed:
modify transaction/urls.py & transaction/views.py

Dm1triiSmirnov committed:
modify transaction/urls.py & transaction/views.py

Dm1triiSmirnov committed:
Merge branch 'master' into transactions_features

# Conflicts:
#	drf_project/settings.py

Dm1triiSmirnov committed:
fix transaction/views.py. Add retrive list of transactions for current user and for given wallet

Dm1triiSmirnov committed:
fix transaction/views.py

Dm1triiSmirnov committed:
add __str__ method to the transaction/models.py

VladKli commented on transaction/serializers.py:
```diff
@@ -15,3 +16,48 @@ class Meta:
             "timestamp",
         )
+
+    def validate(self, data):
```
`def validate()` is better to name validate_currencies

VladKli commented on transaction/serializers.py:
```diff
@@ -15,3 +16,48 @@ class Meta:
+        else:
+            commission = transfer_amount * Decimal("0.1")
+
+        if sender.balance < transfer_amount + commission:
+            raise ValueError("Insufficient balance.")
```
better to create separate validation function that will check that sender has enough balance

something like this, but you also need to add commission 
```
def validate_transfer_amount(self, validated_data):
        sender_wallet = self.get_sender_wallet(validated_data)
        if not sender_wallet.filter(
            balance__gte=Decimal(
                self.context["request"].data["transfer_amount"]
            )
        ).exists():
            raise ValidationError("Not enough balance")

```

Dm1triiSmirnov replied:
Done

VladKli commented on transaction/serializers.py:
you need to add functions that validates sender and receiver wallet (check if the wallets exist)
and add validation function to check if user is owner of a sender wallet

Dm1triiSmirnov replied:
Done

VladKli commented on transaction/serializers.py:
```diff
@@ -15,3 +16,48 @@ class Meta:
+        if sender.balance < transfer_amount + commission:
+            raise ValueError("Insufficient balance.")
+
+        sender.balance -= transfer_amount + commission
+        sender.save()
+        receiver.balance += transfer_amount
+        receiver.save()
```
please check F() expressions for balance transferring

Dm1triiSmirnov replied:
Done

VladKli commented on transaction/views.py:
```diff
@@ -1,9 +1,56 @@
 
 
-class TransactionViewSet(viewsets.ModelViewSet):
+class TransactionListCreateViewSet(
```
post method for transaction creation should be added

VladKli commented on transaction/views.py:
```diff
@@ -1,9 +1,56 @@
+                        status=status.HTTP_404_NOT_FOUND)
+
+
+class TransactionRetrieveDestroyViewSet(
+    mixins.RetrieveModelMixin,
+    mixins.DestroyModelMixin,
+    viewsets.GenericViewSet,
+):
+    queryset = Transaction.objects.all()
+    serializer_class = TransactionSerializer
```
same as with wallets. you need to check if a user is owner

Dm1triiSmirnov replied:
done

Dm1triiSmirnov committed:
add permissions to the views

Dm1triiSmirnov committed:
Merge pull request #2 from Dm1triiSmirnov/wallets

Wallets

Dm1triiSmirnov committed:
Merge branch 'master' into transactions_features

Dm1triiSmirnov committed:
fix/modify transaction:
1. add separate function that check that sender has enough balance
2. add functions that check if the sender/receiver wallets exists
3. add function to check if user is owner of a sender wallet

Dm1triiSmirnov committed:
fix transaction/views.py: add function to check if the user is the owner of transaction

VladKli commented on transaction/serializers.py:
```diff
@@ -15,3 +18,85 @@ class Meta:
+
+        self.check_sender_is_owner(sender)
+
+        if sender.user == receiver.user:
+            commission = 0
+        else:
+            commission = transfer_amount * Decimal("0.1")
+
+        self.validate_sender_balance(validated_data, commission)
+
+        sender.balance = F("balance") - (transfer_amount + commission)
+        sender.save()
+        receiver.balance = F("balance") + transfer_amount
+        receiver.save()
```
please, create separate functions for checking amount of commission and balance transfer 
because now the function is to long. we need to follow SOLID

Dm1triiSmirnov replied:
done

Dm1triiSmirnov committed:
fix transaction: create separate functions for count commision and updating balance
